lib/common/schemas.c

```diff
@@ -1335,3 +1335,111 @@ pcmk__schemas_later_than(const char *name)
 
     return lst;
 }
+
+static void
+external_refs_in_schema(GList **list, const char *contents)
+{
+    xmlNode *ptr = string2xml(contents);
+    /* local-name()= is needed to ignore the xmlns= setting at the top of
+     * the XML file.  Otherwise, the xpath query will always return nothing.
+     */
+    xmlXPathObjectPtr xpathObj = xpath_search(ptr, "//*[local-name()='externalRef'] | //*[local-name()='include']");
+
+    for (int i = 0; i < numXpathResults(xpathObj); i++) {
+        xmlNode *result = getXpathResult(xpathObj, i);
+        const char *href = NULL;
+
+        if (result == NULL) {
+            continue;
+        }
+
+        href = crm_element_value(result, "href");
+
+        if (href != NULL) {
+            char *s = strdup(href);
+
+            CRM_ASSERT(s != NULL);
+            *list = g_list_append(*list, s);
+        }
+    }
+
+    freeXpathObject(xpathObj);
+    free_xml(ptr);
+}
+
+static void
+add_schema_file_to_xml(xmlNode *parent, const char *name, GList **already_included)
+{
+    char *contents = NULL;
+    xmlNode *file_node = NULL;
+    GList *includes = NULL;
+
+    /* If we already included this file, don't do so again. */
+    if (g_list_find_custom(*already_included, name, (GCompareFunc) strcmp) != NULL) {
+        return;
+    }
+
+    contents = pcmk__schema_contents(name);
```

Comment:
I don't know if this would be easier, but `pcmk__schema_contents()` could return an xmlNode* to avoid the `string2xml()` call. `filename2xml()` would probably work.

Reply:
Whatever it returns makes something awkward.

If we return a string, we have to use `string2xml` in `external_refs_in_schema` to do the xpath query.  If we return an xmlNodePtr, we have to convert it to a string for `xmlNewCDataBlock`.